backend/attachRendererFiber.js

```diff
@@ -0,0 +1,244 @@
+      // TODO: the naming is confusing. `element` is *not* a React element. It is an opaque ID.
+      element: fiber._debugID,
+      data: getDataFiber(fiber),
+      renderer: rid,
+      _event: 'mount',
+    });
+    const isRoot = fiber.tag === 3;
+    if (isRoot) {
+      events.push({
+        element: fiber._debugID,
+        renderer: rid,
+        _event: 'root',
+      });
+    }
+  }
+
+  function enqueueUpdate(fiber, events) {
+    events.push({
+      element: fiber._debugID,
+      data: getDataFiber(fiber),
+      renderer: rid,
+      _event: 'update',
+    });
+  }
+
+  function enqueueUnmount(fiber, events) {
+    events.push({
+      element: fiber._debugID,
+      renderer: rid,
+      _event: 'unmount',
+    });
+  }
+
+  function mapChildren(parent, allKeys) {
+    const children = new Map();
+    let node = parent.child;
+    while (node) {
+      const key = node.key || node.index;
+      allKeys.add(key);
+      children.set(key, node);
+      node = node.sibling;
+    }
+    return children;
+  }
+
+  function unmountFiber(fiber, events) {
+    // Depth-first.
+    // Logs unmounting of children first, parents later.
+    let node = fiber;
+    outer: while (true) {
+      if (node.child) {
+        node.child.return = node;
+        node = node.child;
+        continue;
+      }
+      enqueueUnmount(node, events);
+      if (node == fiber) {
+        return;
+      }
+      if (node.sibling) {
+        node.sibling.return = node.return;
+        node = node.sibling;
+        continue;
+      }
+      while (node.return) {
+        node = node.return;
+        enqueueUnmount(node, events);
+        if (node == fiber) {
+          return;
+        }
+        if (node.sibling) {
+          node.sibling.return = node.return;
+          node = node.sibling;
+          continue outer;
+        }
+      }
+      return;
+    }
+  }
+
+  function mountFiber(fiber, events) {
+    // Depth-first.
+    // Logs mounting of children first, parents later.
+    let node = fiber;
+    outer: while (true) {
+      if (node.child) {
+        node.child.return = node;
+        node = node.child;
+        continue;
+      }
+      enqueueMount(node, events);
+      if (node == fiber) {
+        return;
+      }
+      if (node.sibling) {
+        node.sibling.return = node.return;
+        node = node.sibling;
+        continue;
+      }
+      while (node.return) {
+        node = node.return;
+        enqueueMount(node, events);
+        if (node == fiber) {
+          return;
+        }
+        if (node.sibling) {
+          node.sibling.return = node.return;
+          node = node.sibling;
+          continue outer;
+        }
+      }
+      return;
+    }
+  }
+
+  function updateFiber(nextFiber, prevFiber, events) {
+    // TODO: optimize for the common case of children with implcit keys.
+    // Just like we do in the Fiber child reconciler.
+    // We shouldn't be allocating Maps all the time.
```

Comment:
You shouldn't even need to diff on keys and types (which is incomplete anyway) since you can match on object equality between the fiber alternates.

If the first child is equal, then all the children are equal. If the first child is non-equal, you know that you will need to enqueue and update on everything that has an alternate and a mount on anything without an alternate.

The only thing you might need diffing for is unmounts. One solution is to just loop through all the next children with alternates. Add all the alternates to a Set. Then loop through all the previous children, if they're not in the Set, they're deleted.

A better solution might be to just expose a hook for deletions instead.